File: backend-python/totp.py
import base64
import hmac
import hashlib
import time
import struct
import random

def base32_decode(secret):
    secret = secret.strip().upper()
    padding = len(secret) % 8
    if padding:
        secret += '=' * (8 - padding)
    return base64.b32decode(secret)

def hotp(secret_bytes, counter):
    counter_bin = struct.pack('>Q', counter)
    hmac_res = hmac.new(secret_bytes, counter_bin, hashlib.sha1).digest()
    offset = hmac_res[-1] & 0x0f
    code = (
        ((hmac_res[offset] & 0x7f) << 24) |
        ((hmac_res[offset + 1] & 0xff) << 16) |
        ((hmac_res[offset + 2] & 0xff) << 8) |
        (hmac_res[offset + 3] & 0xff)
    )
    return str(code % 1000000).zfill(6)

def verify_totp(secret_base32, token, window=1):
    try:
        secret_bytes = base32_decode(secret_base32)
        current_time = int(time.time())
        time_step = current_time // 30
        
        for i in range(-window, window + 1):
            if hotp(secret_bytes, time_step + i) == token.strip():
                return True
        return False
    except Exception as e:
        print("TOTP verification error:", e)
        return False

def generate_secret(length=32):
    alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567'
    return ''.join(random.choice(alphabet) for _ in range(length))

File: backend-php/totp.php
<?php

class TOTP {
    /**
     * Decodes a base32 string to binary data.
     */
    private static function base32Decode($base32) {
        $alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';
        $base32 = strtoupper(str_replace('=', '', $base32));
        $len = strlen($base32);
        $n = 0;
        $j = 0;
        $binary = '';

        for ($i = 0; $i < $len; $i++) {
            $val = strpos($alphabet, $base32[$i]);
            if ($val === false) {
                throw new Exception('Invalid base32 character');
            }
            $n = ($n << 5) | $val;
            $j += 5;
            if ($j >= 8) {
                $j -= 8;
                $binary .= chr(($n >> $j) & 255);
            }
        }
        return $binary;
    }

    /**
     * Calculates the HOTP code for a secret and counter.
     */
    private static function hotp($secretBinary, $counter) {
        // Pack counter into 8 bytes big-endian (64-bit unsigned)
        $counterBin = pack('N*', 0, $counter);

        // Compute HMAC-SHA1
        $hash = hash_hmac('sha1', $counterBin, $secretBinary, true);

        // Dynamic truncation
        $offset = ord($hash[19]) & 0xf;
        $code = (
            ((ord($hash[$offset]) & 0x7f) << 24) |
            ((ord($hash[$offset + 1]) & 0xff) << 16) |
            ((ord($hash[$offset + 2]) & 0xff) << 8) |
            (ord($hash[$offset + 3]) & 0xff)
        );

        return str_pad($code % 1000000, 6, '0', STR_PAD_LEFT);
    }

    /**
     * Verifies a TOTP token against a base32 secret.
     */
    public static function verify($secretBase32, $token, $window = 1) {
        try {
            $secretBinary = self::base32Decode($secretBase32);
            $timeStep = floor(time() / 30);
            
            for ($i = -$window; $i <= $window; $i++) {
                if (self::hotp($secretBinary, $timeStep + $i) === trim($token)) {
                    return true;
                }
            }
            return false;
        } catch (Exception $e) {
            return false;
        }
    }

    /**
     * Generates a random base32 secret key.
     */
    public static function generateSecret($length = 32) {
        $alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';
        $secret = '';
        for ($i = 0; $i < $length; $i++) {
            $secret .= $alphabet[random_int(0, 31)];
        }
        return $secret;
    }
}
